### ISSUE TITLE
Allow negative learning rates

### ISSUE THREAD

**Issue by yohan-pg** [pytorch/pytorch#46480]
## 🚀 Feature
Currently, optimizers throw an assertion error when negative learning rates are supplied at construction. This proposal suggests removing this restriction. 

## Motivation
Since maximization is equivalent to minimizing a negative loss function, negative learning rates are useful and make sense. For example, in GANs, the generator and discriminator can be trained adversarially by giving the discriminator a negative learning rate. This avoids having two backward passes, which improves both computational efficiency and conceptual clarity. 

## Pitch
As learning rates are typically parameterized by constants, providing a negative rate by accident is highly unlikely. I argue that making this mistake is much less likely than wanting a negative learning rate, and so the defensive assertion is better removed. 

## Alternatives
It is currently possible to set a negative LR through an ugly loop through the optimizer's `param_group`. An alternative would be to provide a cleaner way to do so, i.e., a `maximize=True` flag on optimizer construction. 

cc @ezyang @gchanan @zou3519 @bdhirsh @jbschlosser @anjali411 @vincentqb @iramazanli

**Comment by ezyang** [comment]
bumping priority based on internal request

**Comment by iramazanli** [comment]
I will do related literature survey and start working on this

**Comment by ezyang** [comment]
(iramazanli and I had private correspondence where I told him that literature survey probably not necessary for a patch like this :P)

**Comment by anjali411** [comment]
@albanD verify if just using an alternate approach instead of neg LR would work for the internal use case too

**Comment by albanD** [comment]
From more discussion, I agree that we could add a "maximize=False" argument to all the optimizers and have special logic for each optimizer that should handle this case properly.

**Comment by carlosgmartin** [comment]
@albanD I second @yohan-pg's comment:

> As learning rates are typically parameterized by constants, providing a negative rate by accident is highly unlikely. I argue that making this mistake is much less likely than wanting a negative learning rate, and so the defensive assertion is better removed.

It would be better to simply remove the `lr < 0.0` check from `torch.optim.SGD`'s constructor.

**Comment by OliverFM** [comment]
Hey!
I am new to contributing to torch -- I will open a PR for this issue, if that is alright :)

**Comment by albanD** [comment]
I am not sure everyone agrees yet.
Let me double check with other core contributors on this and update back here.

**Comment by jbschlosser** [comment]
I think a properly functioning `maximize` flag is better than allowing negative learning rates. For optimizers beyond simple SGD, it's likely that negative learning rates will weird undesired behavior. If the goal is to allow maximization, imo this should be provided on a per-optimizer basis to account for the subtleties.

**Comment by yohan-pg** [comment]
> I think a properly functioning `maximize` flag is better than allowing negative learning rates. For optimizers beyond simple SGD, it's likely that negative learning rates will weird undesired behavior. If the goal is to allow maximization, imo this should be provided on a per-optimizer basis to account for the subtleties.

I believe this should be fine with most optimizers? At the very least, this would be nice to have in Adam as it is the most commonly used optimizer. Glancing through the optimizer list on https://pytorch.org/docs/stable/optim.html, I think only Rprop would pose problem, but I am not on expert on optimization by any means (perhaps LBFGS would fail also, although I can't tell).

**Comment by OliverFM** [comment]
> @OliverFM You can go ahead and add the boolean `maximize` argument to the optimizer and update the code to properly do maximization in this case. We want to take extra care to make sure that it works fine with all other arguments and that the doc is correctly updated to reflect the update to the algorithm.

Great -- will do. Since I am also not an optimisation expert, I will start with SGD. Then Adam. Is it important to add support for all?
I am guessing it's better to do this iteratively. What do you think?

**Comment by albanD** [comment]
Iteratively is definitely better. And that's ok to not have all of them yes.

**Comment by albanD** [comment]
Leaving this closed in favor of the tracking issue in https://github.com/pytorch/pytorch/issues/68052
